File: 0128_Longest_Consecutive_Sequence/sol.go
package main

import (
	"fmt"
	"reflect"
)

/**
 * Strategy: Hash set
 * Time:     O(N)
 * Space:    O(N)
 */

func solve(nums []int) int {
	numset := make(map[int]struct{})
	for _, item := range nums {
		numset[item] = struct{}{}
	}
	longstreak := 0
	for x := range numset {
		if _, ok := numset[x-1]; !ok {
			curr_num := x
			curr_streak := 1
			for {
				if _, ok := numset[curr_num+1]; !ok {
					break
				}
				curr_num++
				curr_streak++
			}
			longstreak = max(longstreak, curr_streak)
		}
	}
	return longstreak
}

func main() {
	type testCase struct {
		input1 []int
		expect int
	}
	tests := []testCase{
		{input1: []int{100, 4, 200, 1, 3, 2}, expect: 4},
		{input1: []int{0, 3, 7, 2, 5, 8, 4, 6, 0, 1}, expect: 9},
		{input1: []int{1, 0, 1, 2}, expect: 3},
	}

	for i, tc := range tests {
		got := solve(tc.input1)
		if !reflect.DeepEqual(got, tc.expect) {
			fmt.Printf("❌ Test %d Failed | Input: %v | Got: %v, Want: %v\n",
				i, tc.input1, got, tc.expect)
		} else {
			fmt.Printf("✅ Test %d Passed | Input: %v | Got: %v, Want: %v\n",
				i, tc.input1, got, tc.expect)
		}
	}
}

File: 0128_Longest_Consecutive_Sequence/sol.js
/**
 * @param {number[]} nums
 * @return {number}
 * Time:    O(N)
 * Space:   O(N)
 * Strategy: Hash set
 */
var func = function(nums) {
    let numset = new Set(nums);
    let longstreak = 0;
    for (const x of numset) {
        if(!(numset.has(x-1))){
            let curr_num = x;
            let curr_streak = 1;
            while(numset.has(curr_num+1)) {
                curr_streak++;
                curr_num++;
            }
            longstreak = Math.max(curr_streak, longstreak);
        }
    }
    return longstreak
};
/**
 * Local testing
 */
function test() {
    const testCases = [
        { args: [100,4,200,1,3,2], expected: 4 },
        { args: [0,3,7,2,5,8,4,6,0,1], expected: 9 },
        { args: [1,0,1,2], expected: 3 }
    ];
    testCases.forEach(({ args, expected }, index) => {
        // Use spread syntax (...) to pass array elements as separate arguments
        const result = func(args);
        const status = (result === expected) ? "✅ PASS" : "❌ FAIL";
        console.log(
            `${status} | Test #${index + 1}: ` +
            `Input: ${JSON.stringify(args)} | ` +
            `Expected: ${expected} | ` +
            `Result: ${result}`
        );
    });
}
test();
